Hitting enter while in preview mode will not try to send empty messages

hollywoodno opened the pull request:
When enter is hit in preview mode a validation check is performed using validate() in the compose.js file. There is also finish() which seems also calls validate() and gets me a similar result. I'm trying to figure out which is appropriate to use. 

Tests have not been written. I am working on testing. I have to quickly get re-familiarized with writing the js tests.

Fixes #5574

smarx commented:
*Automated message from Dropbox CLA bot*

@hollywoodno, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
I think `compose.finish` is the correct one to use; that takes care of the related UI updates that we want to make around sending a message.

adnrs96 commented on frontend_tests/node_tests/compose.js:
```diff
@@ -559,6 +561,29 @@ people.add(bob);
+    page_params.enter_sends = false;
+    $("#new_message_content").blur();
+    compose.enter_with_preview_open();
+    assert($("#new_message_content").is_focused());
```
you should probably test the `assert(!send_message_called);` here as well.

adnrs96 commented:
Hi @hollywoodno I took a look at your changes and would like you to ponder a bit upon making use of the `finish` function to fix this. `finish` function is what handles sending message in other normal view I believe.
Doing something like this might be the way to get around the current issue
```
exports.enter_with_preview_open = function () {
    exports.clear_preview_area();
    if (page_params.enter_sends) {
        // If enter_sends is enabled, we just send the message
        exports.finish();
    } else {
        // Otherwise, we return to the compose box and focus it
        $("#new_message_content").focus();
    }
};
```
I would test this out and then make changes to the node tests as required.
What do you think?

showell commented:
@hollywoodno What is the current status of this PR?  I think the node tests here are still kind of verbose.

hollywoodno commented:
@showell I am finish with the PR. In what ways can suggest I improve the node tests and still reach appropriate coverage levels?

showell commented:
Didn't we have some discussion on chat about making the tests more concise?  I think you can stub the `finish` method and just make sure it's called.

hollywoodno commented:
@showell we had some discussion around it. I'm looking over the zulip docs on using stubs in js. I will follow up on chat.

hollywoodno commented:
I will briefly move this to an WIP as I have updated the node tests. It's currently failing because I added additional assertions. If enter_sends is enabled when messages are sent, when they fail to send, it makes sense that enter_sends should still be enabled. I should be done with this later today.

hollywoodno commented:
I modified the node test for test_enter_with_preview open. I didn't stub out finish because both it and validate fails fast. The modifications therefore narrow in on those failures and test for only for the most specific UI updates that were not previously covered .

showell commented:
@hollywoodno  I think your test changes were a good compromise between testing too much and still getting some nice integration checking.  Thanks for your patience on this.  Merged!
